### PR TITLE
Expose flags when allocating memory in tracee.

### PR DESCRIPTION
**Pull request by danielfenner**
When calling mmap in the tracee we previously always allocated the
memory with flags MAP_PRIVATE | MAP_ANONYMOUS.
Now we'll also need to allocate some memory for a stack of a new thread
(context: see bug blow). The flags required for that are:
MAP_ANONYMOUS | MAP_PRIVATE | MAP_STACK | MAP_GROWSDOWN.
We'll keep the MAP_PRIVATE | MAP_ANONYMOUS as a default but expose the
flags to facilitate that use case.

Bug: b/236121842

### CODE REVIEW

File: src/UserSpaceInstrumentation/AllocateInTracee.h [hunk 1]
```diff
@@ -43,11 +44,11 @@ class MemoryInTracee {
   // write permissions. The memory allocated will start at `address`. `address` needs to be aligned
   // to page boundaries. If the memory mapping can not be placed at `address` an error is returned.
   // If `address` is zero the placement of memory will be arbitrary (compare the documentation of
-  // mmap: https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to
-  // the tracee `pid` using `AttachAndStopProcess`.
-  [[nodiscard]] static ErrorMessageOr<std::unique_ptr<MemoryInTracee>> Create(pid_t pid,
-                                                                              uint64_t address,
-                                                                              uint64_t size);
+  // mmap: https://man7.org/linux/man-pages/man2/mmap.2.html). Optionally one can specify the
+  // `flags`passed to mmap.
```
Comment (pierricgimmig) [inline src/UserSpaceInstrumentation/AllocateInTracee.h:48]: nit. add space after `flags`

**Review by pierricgimmig — APPROVED** [review]
Cool, thanks!

**Review by dpallotti — APPROVED** [review]
> The flags required for that are:
MAP_ANONYMOUS | MAP_PRIVATE | MAP_STACK | MAP_GROWSDOWN.

The change looks good but I think you should be aware of the following (you probably already are but to reiterate):

- `MAP_GROWSDOWN` is not necessary if you don't need the stack to grow on demand, i.e., if the initial size is already good enough. Since the new map is going to be used for a specific thread, you might want to set a fixed size for the stack to keep things more consistent.
- `MAP_STACK`: "This flag is currently a no-op on Linux." Then the manpage goes on kind of recommending to use it, but it's not necessary in practice.

In other words, you could get away without this change. Up to you to decide what to do with this.

**Comment by danielfenner** [comment]
> The change looks good but I think you should be aware of the following (you probably already are but to reiterate):

Thanks for clarifying! I was not aware of this. I was under the impression that at `MAP_STACK` is required and `MAP_GROWSDOWN` the usual setting. 

I think we can ignore the compatability considerations for `MAP_STACK`. For `MAP_GROWSDOWN` it turns out pthread_create() produces fixed size stacks for its threads. So I'd actually prefer to go that way. 

As expected things run just fine with a stack allocated with `MAP_PRIVATE | MAP_ANONYMOUS`. I'll just drop this PR.
